_posts/2023-06-04-not-generating-constructors.md

```diff
@@ -0,0 +1,411 @@
+---
+tags:
+- Subspace
+- C++
+- Performance
+- Memory Safety
+---
+
+# Avoiding Death by a Thousand Constructors
+
+**tl;dr We are going to see how Subspace types allow a zero-cost transition from native pointers to
+bounded view types (which means we can get spatial memory safety) that can't be achieved with the
+standard library types.**
+
+Long ago, C and C++ programmers passed around pointers to arrays, hopefully with a size to indicate
+how large the array was in order avoid reading/writing out of bounds, though frequently not. This is
+still a common pattern, though "modern C++" has provided tools to eliminate this class of memory
+safety bug, though the use of "view" types. These are types like `std::string_view` or `std::span`
+in the standard library. Unfortunately, these types don't actually check for using memory out of
+bounds in the C++ standard.
+
+Much has been written about the memory unsafety of lack of bounds checking, and we (as in the
+software industry) have known ways to solve this simple issue for decades now.
+* Here's an article from 1996 (almost 30 years ago)

```

Comment:
If you want, you can go at least as far back as 1975: Protel's 'descriptor' was a dynamically-typed span of N-byte objects. https://noncombatant.org/2023/05/21/protel-sos-dsm-100/ (Obviously with modern templates and generics we can elide the `object_size` field.) CAR Hoare also yelled about this in his 1960s-era Algol implementation in his 1980 Turing Award speech.

This may be my personal bugbear though; it makes me mad that spatial safety was solved before we were born yet we're still fighting for the bare minimum to protect people against governments and organized crime.

Reply:
These are good but not about C++ so also not sure if they fit in right or not. I read.

Reply:
Also yeah big mood.

Reply:
I thought about saying "before C this was already known" but C was 2-4 years before this so I'm not sure how to frame/relate it properly.